rasa/core/policies/ted_policy.py

```diff
@@ -438,186 +479,303 @@ def load(cls, path: Text) -> "TEDPolicy":
+
+        # additional sequence axis is artifact of our RasaModelData creation
+        # TODO check whether this should be solved in data creation
+        x = tf.squeeze(x, axis=1)
+
+        all_labels_embed = self._tf_layers[f"embed.{LABEL}"](x)
 
-        return all_labels, all_labels_embed
+        return all_label_ids, all_labels_embed
 
-    def _emebed_dialogue(self, dialogue_in: tf.Tensor) -> Tuple[tf.Tensor, tf.Tensor]:
+    def _emebed_dialogue(
+        self, dialogue_in: tf.Tensor, sequence_lengths: tf.Tensor
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
         """Create dialogue level embedding and mask."""
 
-        # mask different length sequences
-        # if there is at least one `-1` it should be masked
-        mask = tf.sign(tf.reduce_max(dialogue_in, axis=-1) + 1)
+        mask = self._compute_mask(sequence_lengths)
 
-        dialogue = self._tf_layers[f"ffnn.{DIALOGUE}"](dialogue_in, self._training)
-        dialogue_transformed = self._tf_layers["transformer"](
-            dialogue, 1 - tf.expand_dims(mask, axis=-1), self._training
+        dialogue_transformed = self._tf_layers[f"transformer.{DIALOGUE}"](
+            dialogue_in, 1 - mask, self._training
         )
         dialogue_transformed = tfa.activations.gelu(dialogue_transformed)
 
         if self.max_history_tracker_featurizer_used:
-            # pick last label if max history featurizer is used
-            dialogue_transformed = dialogue_transformed[:, -1:, :]
-            mask = mask[:, -1:]
+            # pick last vector if max history featurizer is used
+            dialogue_transformed = tf.expand_dims(
+                self._last_token(dialogue_transformed, sequence_lengths), 1
+            )
+            mask = tf.expand_dims(self._last_token(mask, sequence_lengths), 1)
 
         dialogue_embed = self._tf_layers[f"embed.{DIALOGUE}"](dialogue_transformed)
 
         return dialogue_embed, mask
 
-    def _embed_label(self, label_in: Union[tf.Tensor, np.ndarray]) -> tf.Tensor:
-        label = self._tf_layers[f"ffnn.{LABEL}"](label_in, self._training)
-        return self._tf_layers[f"embed.{LABEL}"](label)
+    def _encode_features_per_attribute(
+        self, tf_batch_data: Dict[Text, Dict[Text, List[tf.Tensor]]], attribute: Text
+    ) -> Optional[tf.Tensor]:
+
+        if not tf_batch_data[attribute]:
+            return None
+
+        attribute_mask = tf_batch_data[attribute][MASK][0]
+        # TODO transformer has to be used to process sequence features
+        attribute_features = self._combine_sparse_dense_features(
+            tf_batch_data[attribute][SENTENCE],
+            f"{attribute}_{SENTENCE}",
+            mask=attribute_mask,
+        )
+
+        if attribute in FEATURES_TO_ENCODE + LABEL_FEATURES_TO_ENCODE:
+            attribute_features = self._tf_layers[f"ffnn.{attribute}_{SENTENCE}"](
+                attribute_features
+            )
+
+        return attribute_features * attribute_mask
+
+    def _process_batch_data(
```

Comment:
can we please add a comment to this method explaining what it does? otherwise I think it is a bit hard to understand

Reply:
done!